### PR TITLE
gh-101293: Fix support of custom callables and types in inspect.Signature.from_callable()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
Support callables with the `__call__()` method and types with `__new__()` and `__init__()` methods set to class methods, static methods, bound methods, partial functions, and other types of methods and descriptors.

Add tests for numerous types of callables and descriptors.

* Issue: gh-101293

### CODE REVIEW

File: Lib/inspect.py [hunk 1]
```diff
@@ -2533,7 +2542,8 @@ def _signature_from_callable(obj, *,
         # Unwrap until we find an explicit signature or a MethodType (which will be
         # handled explicitly below).
         obj = unwrap(obj, stop=(lambda f: hasattr(f, "__signature__")
-                                or isinstance(f, types.MethodType)))
+                                or isinstance(f, types.MethodType)
+                                or not callable(f.__wrapped__)))
```
Comment (gaogaotiantian) [inline Lib/inspect.py]: In what scenario, is `f.__wrapped__` not `callable`? It might worth some comments as it's not easy to follow.

File: Lib/inspect.py [hunk 2]
```diff
@@ -2609,28 +2619,27 @@ def _signature_from_callable(obj, *,
...
-        call = _signature_get_user_defined_method(type(obj), '__call__')
-        if call is not None:
-            sig = _get_signature_of(call)
+        call = getattr_static(type(obj), '__call__', None)
+        if call is not None and not isinstance(call, _NonUserDefinedCallables):
+            call = _descriptor_get(call, obj)
+            return _get_signature_of(call)
         else:
-            factory_method = None
             new = _signature_get_user_defined_method(obj, '__new__')
```
Comment (gaogaotiantian) [inline Lib/inspect.py]: It seems like `__new__` is the only exemption that continues to use this original method. All the others are replaced with the new logic which uses `getattr_static` to avoid descriptor resolving. Is there a reason for that? Can we simply change `_signature_get_user_defined_method`?

File: Lib/inspect.py [hunk 3]
```diff
@@ -2673,13 +2682,10 @@ def _signature_from_callable(obj, *,
...
-        call = _signature_get_user_defined_method(type(obj), '__call__')
+        call = getattr_static(type(obj), '__call__', None)
         if call is not None:
-            try:
-                sig = _get_signature_of(call)
-            except ValueError as ex:
-                msg = 'no signature found for {!r}'.format(obj)
-                raise ValueError(msg) from ex
+            call = _descriptor_get(call, obj)
+            return _get_signature_of(call)
```
Comment (gaogaotiantian) [inline Lib/inspect.py:2679]: Did we get rid of this `ValueError` with the new logic?

File: Lib/test/test_inspect/test_inspect.py [hunk 4]
```diff
@@ -3484,6 +3495,98 @@ def __init__(self, b):
...
+
+        with self.subTest('BuiltinMethodType'):
+            class CM(type):
+                __call__ = ':'.join
+            class C(metaclass=CM):
+                def __init__(self, b):
+                    pass
+
+            self.assertEqual(C(['a', 'bc']), 'a:bc')
+            # BUG: Returns '<Signature (b)>'
```
Comment (gaogaotiantian) [inline Lib/test/test_inspect/test_inspect.py:3602]: I have some doubts listing the bug as a test case. If we know it's a wrong answer, we should not test that the answer is accurately generated. It's possible in the future something changed and this *wrong* result is corrected, or converted to another wrong one, and the test would fail - and people will be confused unless they look at the source test code and realize it was labeled "BUG". It may help for the "backward compatibility" if we consider the bug as a compatibility we need to keep track of, but I still have concerns for this. If we do want to keep track of it, I think a `#TODO` would be much better.

File: Lib/test/test_inspect/test_inspect.py [hunk 5]
```diff
@@ -3545,6 +3648,169 @@ def __init__(self, b):
                            ('bar', 2, ..., "keyword_only")),
                           ...))
 
+    def test_signature_on_class_with_init(self):
+        class C:
+            def __init__(self, b):
+                pass
+
+        C(1)
```
Comment (gaogaotiantian) [inline Lib/test/test_inspect/test_inspect.py]: I understand that this is to test whether the method works without an exception, but this statement is a bit weird by itself. What are the possible use cases for a `@staticmethod` `__init__`? To be honest all the decorators on `__init__` and `__new__` are so weird that I doubt if there's anyone actually using them.

File: Lib/inspect.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py]: When it is a data descriptor. For example `staticmethod.__wrapped__`.

It is actually a workaround of issue #112006. We need `obj` to be a callable, so we can use this workaround here. But more general solution can replace it.

File: Lib/inspect.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py]: It reproduces the logic in the C code (see `slot_tp_call`, `slot_tp_new` and `slot_tp_init`). `getattr_static()` should be used for `__call__` and `__init__`, and `getattr()` for `__new__`.

File: Lib/inspect.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py:2679]: Previously the error for invalid signature
```py
def __call__(*, a):
```
was raised outside of the try/except block, in `_signature_bound_method()`.

But now the logic is changed, and this error is raised in `_get_signature_of(call)`. We do not want to replace it with more generic error "no signature found" (there is an existing test for specific error message).

File: Lib/test/test_inspect/test_inspect.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/test/test_inspect/test_inspect.py:3602]: This is definitely a bug. The current code ignores `__call__` signature and gets the signature from `__init__`, which is not even called. I used label TODO if `signature()` currently raises ValueError, but can get a right result, and label BUG if it successfully returns wrong result. The test is added to show what the right answer is. If in the future something changed and this wrong result is corrected, it will be good, we will remove `assertRaises()` and the comment.

We can also split these tests on small methods and use the `@expectedFailure` decorator, if it matters.

BTW, I have a patch that fixes these TODOs and BUGs, but I am not certain about it, so I kept only safer changes here.

File: Lib/test/test_inspect/test_inspect.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/test/test_inspect/test_inspect.py]: They are weird because they are the only examples that can be created using existing builtin classes. But in future we could implement `__init__`, `__new__` and `__call__` in C for some classes. By testing with different types of callables and descriptors we increase confidence in working with arbitrary appropriate types.

Comment (gaogaotiantian) [reply Lib/test/test_inspect/test_inspect.py]: For the first sentence I actually meant the `C(1)` statement - it does not explicitly check against anything. I understand that it tests whether an exception will be raised when instantiating the object, but it still looks a bit weird.

File: Lib/inspect.py (hunk 1 above)
Comment (gaogaotiantian) [reply Lib/inspect.py]: Okay sure that makes sense. I still believe that worth some comments about it (the comments above covered the other cases listed and they are much easier to follow than `f.__wrapped__)`

File: Lib/inspect.py (hunk 2 above)
Comment (gaogaotiantian) [reply Lib/inspect.py]: I think `_signature_get_user_defined_method` is still a pretty self-explanatory name for the method. The only usage for this private method is to get user defined `__new__`, `__init__` and `__call__` methods. Can we do the check logic inside the method, asserting the input method to check, and use `getattr()` / `getattr_static()` accordingly? The only difference between `__new__` and the other two is which function to use right? We don't even need the `_descriptor_get` function, and we don't need to change the logic in this function.

File: Lib/inspect.py (hunk 3 above)
Comment (gaogaotiantian) [reply Lib/inspect.py:2679]: So when did this removed exception triggered before the change? Because at least from the search for the exception message, there are no other such error messages in this file, which means an error case that triggers this message before, now either triggers other messages, not does not trigger exception at all.

File: Lib/test/test_inspect/test_inspect.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/test/test_inspect/test_inspect.py]: I'll add comments.

File: Lib/inspect.py (hunk 3 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py:2679]: It was triggered in the example
```py
class ThisWorksNow:
    __call__ = type
inspect.signature(ThisWorksNow())
```
Previously it raised a ValueError with message "no signature found for <...ThisWorksNow object at ...>". Now it raises a ValueError with message "no signature found for builtin <class 'type'>". The test only checks for "no signature found", so it is passed. The fact that the test was so lenient can mean that it does not matter.

Note that all other recursive calls of `_get_signature_of()` are not wrapped in try/except blocks, so the current code is inconsistent in wrapping exceptions. If it is desirable, it should be done at the upper level, in `Signature.from_callable()` (`getfullargspec()` already wraps any exception into a TypeError).

File: Lib/inspect.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py]: I'll add a comment.

File: Lib/inspect.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py]: We can do this, but it will split the logic on two parts, 600 lines apart one from other. There are other differences:

* Note that the code for `__new__` is followed by `break` (to call `_signature_bound_method()` if `skip_bound_arg` is true), while the code for other methods immediately returns. So some logic will be here, and some in `_signature_get_user_defined_method()`.

* The code for non-type's `__call__` does not have restrictions for builtin types. In future, I want to loosen restrictions for other methods too, so `isinstance(meth, _NonUserDefinedCallables)` will be different for every method, and we may need to check the type of the descriptor before resolving it.

File: Lib/inspect.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py]: This workaround no longer needed.

**Comment by serhiy-storchaka** [comment]
@gaogaotiantian, which of my comments answered your concerns and which ones were unconvincing?

File: Lib/inspect.py (hunk 2 above)
Comment (gaogaotiantian) [reply Lib/inspect.py]: Okay I still think the logic is hard to follow. It took me a while to figure out that the key thing changed is that `init` and `call` are currently bound methods, so they do not need the strip of the first parameter - this logic for `new` however, is far down.

I still believe that we should put the new descriptor/getattr_static logic into `_signature_bound_method()` because that's exactly what they do. You still have the old function and that's having two super similar routine far away from each other.

I think the logic in `_signature_from_callable` should be as simple as

```python
call = _signature_get_user_defined_method(type(obj), '__call__')
new = _signature_get_user_defined_method(obj, '__new__')
init = _signature_get_user_defined_method(obj, '__init__')
```

However, we need to deal with the bound-issue, which is what `slot_tp_call` does - we can just return another indicator:

```python
call, unbound = _signature_get_user_defined_method(type(obj), '__call__')
```

We would significantly simplify the logic in `_signature_from_callable` and get what we need (function and whether it's bounded) with a single function call.

Then, inside `isinstance(obj, type)` check, before `if sig is None:`, we can simply do

```python
if sig is not None:
    if unbound and skip_bound_arg:
        return _signature_bound_method(sig)
    return sig
```

Then of course the following code does not need the if check protection as we ensured `sig is not None` at that point.

This way, we have all the related logic in two places:

1. `_signature_get_user_defined_method` that "magically" does the check and resolves the descriptor.
2. A continuous block in `if isinstance(obj, type):` that solves everything.

Comment (serhiy-storchaka) [reply Lib/inspect.py]: My goal was to minimize the diffs for easier review and do refactoring in the following PR. But if you prefer, I do some refactoring in this PR (complete refactoring will be possible after making some other behavior changes).

I tried to implement your suggestions, except returning `unbound`, because I do not see how it can help.

File: Lib/inspect.py [hunk 6]
```diff
@@ -2600,96 +2611,72 @@ def _signature_from_callable(obj, *,
...
+                raise ValueError(
+                    'no signature found for builtin type {!r}'.format(obj))
 
-    if isinstance(obj, types.BuiltinFunctionType):
-        # Raise a nicer error message for builtins
-        msg = 'no signature found for builtin function {!r}'.format(obj)
-        raise ValueError(msg)
+    else:
+        # An object with __call__
+        call = getattr_static(type(obj), '__call__', None)
```
Comment (gaogaotiantian) [inline Lib/inspect.py:2676]: So this can deal with non user defined methods? And that's why it's not using `_signature_get_user_defined_method(obj, '__call__')`? And that would be fixed in the future? I mean personally, I would just pass another (optional) argument to `_signature_get_user_defined_method` for whether to allow non user defined methods, but if this is going to be fixed soon, I'm okay for this being here temporarily.

File: Lib/inspect.py [hunk 7]
```diff
@@ -2492,6 +2494,15 @@ def _signature_from_function(cls, func, skip_bound_arg=True,
                __validate_parameters__=is_duck_function)
 
 
+def _descriptor_get(descriptor, obj):
```
Comment (gaogaotiantian) [inline Lib/inspect.py:2497]: Would it make sense to move this function closer to `_signature_get_user_defined_method`? Or do you think it belongs here better?

Comment (serhiy-storchaka) [reply Lib/inspect.py:2497]: I added it close to `_signature_from_callable()` where it was used. Now it is also used in `_signature_get_user_defined_method()` which I planned to remove. If move it close to `_signature_get_user_defined_method()`, it will be far from `_signature_from_callable()`.

File: Lib/inspect.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py:2676]: `_signature_get_user_defined_method` is more than 600 lines away from this place. You need to scroll 25 pages up and down to follow the logic.

Comment (gaogaotiantian) [reply Lib/inspect.py:2676]: That's simply the case for almost all function calls. I don't think how far away the function is should be the deterministic factor for whether to use a function - otherwise we should just inline all the code. From my perspective, the "readability" of the code is how well the readers can understand the purpose of the code, and `_signature_get_user_defined_method` is explanatory.

So the question really is - is this trying to get the user defined method? Why it was different (not checking `_NonUserDefinedCallables`). Is it because the `_NonUserDefinedCallables` actually works in this case, or it's impossible to get `_NonUserDefinedCallables`(so it's actually okay to directly call `_signature_get_user_defined_method`)? 

I wouldn't have a clue why `getattr_static` is used or why we need to do a `_descriptor_get` without some deep knowledge - basically I would've lost here. But if it's a function `_signature_get_user_defined_method`, without knowing the detail of the implemenentation, I would at least understand - oh you are getting user defined method here, and that's often enough for reading the code here. That's why I wanted to put these logic into a function. In the future, if someone realize that a specific logic should be changed here, they would not need to search all the `getattr_static` usage, which probably scatters around, they only need to work in the only function.

Also, for any modern IDEs, 600 lines away is just a simple click - I don't think anyone is scrolling their screen to follow the function anymore. In VSCode, it's `Ctrl+click` to follow the function, `Alt+left arrow` to go back to the caller. Even with vim, that's a quick search in the same file.

File: Lib/inspect.py (hunk 7 above)
Comment (gaogaotiantian) [reply Lib/inspect.py:2497]: You plan to move `_signature_get_user_defined_method()` or the usage of `_descriptor_get` in `_signature_get_user_defined_method()`? I was expecting for removal of `_descriptor_get` in `_signature_from_callable()` but maybe we are look at different paths.

Comment (serhiy-storchaka) [reply Lib/inspect.py:2497]: I plan to remove `_signature_get_user_defined_method()`. I only left it to minimize the diff.

Comment (gaogaotiantian) [reply Lib/inspect.py:2497]: How do you plan to replace all the `_signature_get_user_defined_method()` calls for `__call__`, `__init__` and `__new__`? Are you going to do a

```python
meth = getattr(cls, '__new__', None)
meth = _descriptor_get(meth, cls)
```

when the `_NonUserDefinedCallables` issue is solved?

File: Lib/inspect.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py:2676]: The _NonUserDefinedCallables actually works in this case, and the main purpose of this PR is to support signature when `__call__` is _NonUserDefinedCallables.

It actually mostly works in other three cases, but with caveats, different for different cases, so I prefer to leave this for other PR. This is why I plan to replace complex `_signature_get_user_defined_method` (its name and description will no longer be correct) and replace it with simpler 2-4 lines for every concrete case.

File: Lib/inspect.py (hunk 7 above)
Comment (serhiy-storchaka) [reply Lib/inspect.py:2497]: No need to use `_descriptor_get` for `__new__`. In this case it can be just one line.

File: Lib/inspect.py (hunk 6 above)
Comment (gaogaotiantian) [reply Lib/inspect.py:2676]: Hmm okay, that makes sense. So if the function is not getting user defined method anymore (that it can take system-defined functions), we should at least change it. I guess we'll see if it's better to just remove the function and do a couple of lines for each usage, or to create another function with a explanatory name so the user would know that all these places are doing similar things.

**Review by gaogaotiantian — APPROVED** [review]

**Comment by gaogaotiantian** [comment]
Thank you for your patience. I approved the change for now because obviously you have future plans with the code and some of them will be refactored soon, so it's not worth it to debate on the best shape of the current situation. It's completely acceptable now.

However, as I'm not the core-dev so you probably need another person to take a look at the code?

**Comment by serhiy-storchaka** [comment]
Thank you for your review @gaogaotiantian.
